harp/reader.py

```diff
@@ -44,13 +44,53 @@ def __init__(
         self.read = read
 
 
+class RegisterMap:
+    registers: dict[str, RegisterReader]
+    _register_map: dict[int, str]
+
+    def __init__(self, registers: dict[str, RegisterReader]) -> None:
+        self.registers = registers
+        self._register_map = {self.registers[key].register.address: key for key in self.registers.keys()}
+
+    def __getitem__(self, __key: Union[str, int]) -> RegisterReader:
+        if isinstance(__key, int):
+            return self.registers[self._register_map[__key]]
+        elif isinstance(__key, str):
+            return self.registers[__key]
+        else:
+            raise TypeError(f"key must be int or str, not {type(__key)}")
+
+    def __iter__(self) -> Iterable[str]:
+        return iter(self.registers)
+
+    def __len__(self) -> int:
+        return len(self.registers)
+
+    def __repr__(self) -> str:
+        return repr(self.registers)
+
+    def __str__(self) -> str:
+        return str(self.registers)
+
+    def __dir__(self) -> Iterable[str]:
+        return self.registers.keys()
+
+    def __getattr__(self, __name: str) -> RegisterReader:
+        return self.registers[__name]
+
+
 class DeviceReader:
     device: Model
-    registers: dict[str, RegisterReader]
+    registers: RegisterMap
 
-    def __init__(self, device: Model, registers: dict[str, RegisterReader]) -> None:
+    def __init__(self, device: Model, registers: Union[RegisterMap, dict[str, RegisterReader]]) -> None:
         self.device = device
-        self.registers = registers
+        if isinstance(registers, RegisterMap):
+            self.registers = registers.registers
```

Comment:
In this case, shouldn't we just assign the `registers` object itself, rather than the inner map?

Reply:
Yup, my mistake!